### DISCUSSION TITLE
how to show logger warnings and info

### DISCUSSION THREAD

**Discussion by grgmiller** [catalyst-cooperative/pudl#2298] [Help Me!] [answered]
When calling the `pudl.analysis.allocate_net_gen` functions as part of our OGE pipeline, I'm wondering how we can get the logger warning and info messages to show up. Generally, whenever I run the `allocate_net_gen` pipeline, whether in a jupyter notebook, or in the cmd line, I never see any messages and I'm realizing that I'm probably missing important warnings. I know this is a pretty basic question but I don't have a lot of experience using `logging`. In my particular situation, I have pudl installed as part of a conda environment and would typically be running something like:
```
import pudl.analysis.allocate_net_gen

pudl_out = load_data.initialize_pudl_out(year=year)

gen_fuel_allocated = pudl.analysis.allocate_net_gen.allocate_gen_fuel_by_generator_energy_source(
    pudl_out, drop_interim_cols=True
)
```

**Comment by zaneselvans** [comment]
In a conda environment with `catalystcoop.pudl` installed, the following generates logging output for me and takes a few seconds to run. Does it produce any logging output for you? 

```python
import pudl
import sqlalchemy as sa

pudl_settings = pudl.workspace.setup.get_defaults()
pudl_engine = sa.create_engine(pudl_settings["pudl_db"])
pudl_out = pudl.output.pudltabl.PudlTabl(pudl_engine)
gf_eia923 = pudl_out.gf_eia923()
```

This gave me logging output in a Jupyter notebook, in IPython at the console, and at a traditional python prompt. If it's not producing output for you, maybe your Python logging environment is configured differently? Are you configuring logging anywhere in the OGE package / environment?

### Complicated stuff

It took me a long time to wrap my head around the python `logging` module, and I definitely still don't understand it completely.

Generally, you don't want all logging from every imported package/module to show up on your console, since you could be using code from hundreds of packages, many of which you have no control over. Well-behaved packages only tell their internal loggers to generate output that you see if asked for it. The logging system has 3 main components:
* loggers (the objects that generate logging messages)
* formatters (which determine how the logging messages are displayed and what metadata you see)
* handlers (which tell the logging messages where they should go -- the console, stderr, a file, etc.).

A confusing (but useful) part of the logging system is that there's a hierarchy of loggers, all of which feed their logging messages up to a root logger, that typically takes care of all of the output (because it's typically the only logger with a Handler attached).

If you create a logger with a scope that includes PUDL (or the module you care about getting logging messages from) it will automatically be a child of the root logger, and should, I think give you the output you want to see (assuming you're getting other logging messages already).

```python
# This will create a logger that's listening for log messages from all the PUDL sub-packages and modules:
pudl_logger = logging.getLogger(name="catalystcoop.pudl")
```

If you don't explicitly attach a Handler to the root logger, I think it's done automatically, but you can either do it explicitly with the `Logger.addHandler()` method or by calling [`logging.basicConfig()`](https://docs.python.org/3/library/logging.html#logging.basicConfig).

**Comment by grgmiller** [comment]
So if I run the code snippet you shared above in a jupyter notebook, I am seeing the logging output. 

However, if I run the following:

```
import pudl.analysis.allocate_net_gen as allocate_gen_fuel

pudl_out = load_data.initialize_pudl_out(year=year)

# allocate net generation and heat input to each generator-fuel grouping
gen_fuel_allocated = allocate_gen_fuel.allocate_gen_fuel_by_generator_energy_source(
    pudl_out, drop_interim_cols=True
)
```

I do not see any logging output.

I also tried running

```
import pudl.analysis.allocate_net_gen as allocate_gen_fuel
import logging

pudl_logger = logging.getLogger(name="catalystcoop.pudl")

pudl_out = load_data.initialize_pudl_out(year=year)

# allocate net generation and heat input to each generator-fuel grouping
gen_fuel_allocated = allocate_gen_fuel.allocate_gen_fuel_by_generator_energy_source(
    pudl_out, drop_interim_cols=True
)
```
and did not see any logging output. This seems strange to me since the `pudl.analysis.allocate_net_gen` module also has logger outputs that should probably be showing.

Any ideas on why these logging outputs might not be showing up when I run a higher-level function?

**Reply by zaneselvans** [reply]
Try adding a stand alone `import pudl`.  The `allocate_net_gen` has logging output in it, but if the PUDL logger hasn't been created and attached to a handler (which I think is done in the top-level `pudl.__init__.py` then the outputs won't go anywhere.

What you do below -- explicitly creating the `pudl_logger` serves the same purpose (I could be wrong, but I think it would also show logging output even if you didn't set the logging level)

**Comment by grgmiller** [comment]
I figured out what needed to happen- I need to manually set the logging level. The following code now works:

```
import pudl.analysis.allocate_net_gen as allocate_gen_fuel
import logging

pudl_logger = logging.getLogger(name="catalystcoop.pudl")
pudl_logger.setLevel("INFO")

pudl_out = load_data.initialize_pudl_out(year=year)

# allocate net generation and heat input to each generator-fuel grouping
gen_fuel_allocated = allocate_gen_fuel.allocate_gen_fuel_by_generator_energy_source(
    pudl_out, drop_interim_cols=True

```

**Reply by grgmiller** [reply]
Well... this works for jupyter notebooks, but when I try and run OGE through the command line, the logger output is still not showing

**Reply by zaneselvans** [reply]
Can you create a logger in the script that you're using to run OGE, and run `basicConfig()` to make sure that the root logger has a handler attached?

@jdangerx @zschira @bendnorman does any of this sound sensible to you?

**Comment by miloknowles** [comment, marked as answer]
Hi all, we discovered the reason PUDL's logs were disappearing when we ran the OGE pipeline in case anyone is curious. Another package we import applies a configuration file to the root logger using this function:

```python
logging.config.fileConfig(LOG_CONFIG_FILE_PATH)
```

Looking at the [docs](https://docs.python.org/3/library/logging.config.html#logging.config.fileConfig), the signature for this function is:
```python
logging.config.fileConfig(fname, defaults=None, disable_existing_loggers=True, encoding=None)
```

For some reason, `disable_existing_loggers=True` is the default, so anyone who calls this function on the root logger will turn off everyone else's loggers that have been configured already. After turning the flag off, we can see PUDL's logs.

Thanks for your help!

**Reply by zaneselvans** [reply]
Wow, that's frustrating! Glad you figured it out, and thanks for bringing the answer back here too!
